### ISSUE TITLE
What does "accuracy per sequence" mean and how BLEU score is computed?

### ISSUE THREAD

**Issue by CaoKaibo** [tensorflow/tensor2tensor#1804]
### Description
Hello, I have a question during my using t2t for a text2text problem.
I defined a problem and used my own dataset for training, the tensorboard shows the evaluation metrics, which includes the accuracy, accuracy per sequence, and the BLEU score.

My question is
1. what's the difference between "accuracy" and "accuracy per sequence"
2. How is BLEU computed? (For a sequence or a token)

Let's assume that my input is
`input1, input2, input3`
and the output is
`output1, output2, output3, output4`
and the ground truth is
`truth1, output2, truth3`

so, for this training example, what is the score of "accuracy" and "accuracy per sequence"? After reading the `metrics.py`, I think the "accuracy" is calculated for each token? (1/3 for this example since the output2 is predicted correctly), and "accuracy per sequence" is 0? (because only when the output and ground truth is completely identical (allow for the incorrect of output4) the value can be 1, and the "accuracy per sequence" is calculated as the `completely_matched_example`/`all_example`). Is this understanding correct?

For question 2, it is similar to question 1, when calculating the BLEU score, it is calculated for each example and calculate an average for all examples or it is calculated for each token and discard the concept of the "examples" (i.e. input a sentence and output a sentence).

**Comment by martinpopel** [comment]
1. "accuracy" is the percentage of correctly predicted tokens, "accuracy per sequence" is the percentage of correctly predicted sequences (sentences), i.e. sequences where all tokens are correctly predicted.
By default, the T2T evaluation is not autoregressive, ie. it works as in training - the task is just to predict the next token, given all previous tokens are predicted correctly (of course, this is not realistic and it underestimates the error, but it is much faster to compute). So it cannot happen that the predicted output has 4 tokens, but the ground truth just 3 tokens as in your example. This could happen only with `--eval_run_autoregressive` or in the real decoding (using `t2t-decoder`, where also the beam search can be taken into consideration).

2. See the [original paper about BLEU](https://www.aclweb.org/anthology/P02-1040.pdf) for its definition. BLEU is a document-level metric and cannot be decomposed as a function (e.g. weighted average) of sentence-level BLEU scores (nor any token-level scores). The reason is the BLEU brevity penalty.

**Comment by CaoKaibo** [comment]
@martinpopel Thank you very much for your reply.

For question 1, I have understood. But for question 2, after reading the paper, I still have some questions.

Let‘s keep using my example mentioned above, for such an example, we can calculate the BLEU score according to the model output and ground truth. (If I understand correctly, because the model only outputs these few words, these few words (this sentence) are what you call the **document**-level).

And my question is how is the final score (t2t-bleu.py returned) is calculated? My validation/test set contains many samples like the example I mentioned above, so how is the final BLEU result calculated? Is it an average of all samples?

**Comment by martinpopel** [comment]
If your document (test set) contains only a single sentence, you can compute BLEU on it, but it will be very unreliable. Averaging e.g. 1000 sentence-level BLEU scores gives a different value (and much less reliable) than computing one document-level BLEU on the same 1000-sentence test set. By reliability I mean correlation with human judgement. If you need sentence-level MT evaluation, use e.g. chrF score (which is also character level, which may be another advantage for some languages).

For understanding BLEU, it may be easier to read [the source code](https://github.com/tensorflow/tensor2tensor/blob/598eace7707d77b7c5117a2235ac3c4924796ca2/tensor2tensor/utils/bleu_hook.py#L60-L129) rather than the original paper. Note that `for (references, translations) in zip(reference_corpus, translation_corpus):` means for each sentence (each pair of system translation and the respective reference translation) in the test set.

Why I called BLEU a document-level metric?
Note that BLEU = brevity_penalty * geometric_mean(precision of unigrams...4grams). If your test set contains a single sentence only and if there is no matching 4gram (which can happen in a single sentence), the 4gram precision is 0 and thus the whole BLEU should be 0. The implementation in T2T uses "smoothing", so the final BLEU will be non-zero, but still very low and unreliable. Another reason is the brevity penalty (BP): if the single sentence has e.g. 4 tokens and the reference 6 tokens, BP=exp(1-6/4)=0.6, so the BLEU score is just 60% of the geometric mean, which seems as a too harsh penalty for being only two tokens shorter than the reference. On the other hand if the single sentence has 6 tokens and the reference 4, the unigram precision can be at most 4/6 (if all 4 tokens are matching the reference), the bigram precision at most 3/5, trigram 2/4 and 4gram 1/3, so the geometric mean (and the final BLEU) will be at most 0.508, which seems again as a too harsh punishment for adding only two extra tokens.

**Comment by CaoKaibo** [comment]
@martinpopel Thanks again for your detailed answer.

> If your document (test set) contains only a single sentence, you can compute BLEU on it, but it will be very unreliable. If you need sentence-level MT evaluation, use e.g. chrF score.

I will change the calculation method to document-level and try to use chrF (my corpus is English-only).

> For understanding BLEU, it may be easier to read the source code rather than the original paper.

After reading the [the source code of BLEU](https://github.com/tensorflow/tensor2tensor/blob/598eace7707d77b7c5117a2235ac3c4924796ca2/tensor2tensor/utils/bleu_hook.py#L60-L129), I found that it calculates the final score by first calculating and adding n-gram overlap for all the sentences in the test set together and calculate the final score for one time instead of doing sentence-level calculation and then doing average. Is this correct?

> If your test set contains a single sentence only and if there is no matching 4gram (which can happen in a single sentence), the 4gram precision is 0 and thus the whole BLEU should be 0.

In my experiment, the model will only output a few words, about 2-6. So when calculating the BLEU score, n-gram overlap maybe 0 (4-gram overlap for sentence whose length is 2). For this situation, my previous approach was to set N in the N-gram to max(length_model_output, length_ground_truth) by setting weights, if it is greater than 4, then N is set to 4.

Then I used the sentence-level calculation method to calculate the BLEU score for each sample in the test set separately, and finally take the average value. 

I don't know if this is a common practice? What are the disadvantages of doing like this?
(As you said, BLEU score is a document-level metric, so it is better to use document-level calculation method. Doing so will definitely cause some unreliability.)

Another question is if I use the document-level calculation, how to choose the parameter N, since the length of the sentence output by the model is very short (mostly 2-4).

**Comment by martinpopel** [comment]
> I will change the calculation method to document-level and try to use chrF (my corpus is English-only).

Note that chrF is not available in T2T, but e.g. in [sacreBLEU](https://github.com/mjpost/sacreBLEU).

> Is this correct?

Yes. That's the whole point.

> In my experiment, the model will only output a few words, about 2-6.

Then BLEU is not suitable for this task. Restricting n-gram from 4 to 1 or 2 should help a bit, but the suitability is still questionable.
It may be wise to look if someone else does this type of task and what metric they used.
It depends on the task - e.g. when predicting names of cities and the reference answer is "New York", then "New Orleans" is as bad answer as "Miami" despite the one word overlap. In this case, you are probably interested in exact match (accuracy_per_sequence).

**Comment by CaoKaibo** [comment]
@martinpopel Thanks a lot!

I do use exact match as my evaluation metric. In order to make the results more comprehensive, I am also considering other metrics. I used BLEU and ROUGE before, now, I might turn to others...
